### PR TITLE
For Mentor Review | Output And Error Handling | Baz Murphy

### PR DESCRIPTION
**Pull request by bazmurphy**
# Learners, PR Template

## Self checklist

- [x] I have titled my PR with `For Mentor Review | {Project Name} | {My Name}`
- [ ] I have included a list of changes **(initial commit)**
- [ ] I have tested my changes **(have not yet implemented any tests)**
- [ ] I have explained my changes **(initial commit)**
- [x] I have asked someone to review my changes

## Description

This is my first attempt at project 01 `output-and-error-handling`

### Change list
Not applicable (initial commit)

## Rationale, briefly

- What is the goal of these changes?
Not applicable (initial commit)

- How did these changes address this goal or goals?
Not applicable (initial commit)

- Why are they a good idea?
Not applicable (initial commit)

## @mentions, #links

- Who needs to know about this? @mention them
@illicitonion Daniel

- Is this related to another issue, PR, or repo? #link them
Not applicable (initial commit)

### CODE REVIEW

File: projects/output-and-error-handling/client/main.go [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+		fmt.Fprintln(os.Stderr, "Parsing Response Body Failed.")
+		return
+	}
+
+	// deliberately using os.Stdout (not fmt.Println) even though unnecessary (for the sake of the exercise)
+	fmt.Fprintln(os.Stdout, string(responseBody))
+}
+
+// note: returns an integer: 0 for fail || > 0 for success
+func parseRetryAfterHeader(retryAfterHeaderString string) int {
```
Comment (berkeli) [inline projects/output-and-error-handling/client/main.go:36]: Can you think of a way to re-implement this function using the examples for [error propagation](https://github.com/bazmurphy/immersive-go-course/tree/main/projects/output-and-error-handling#1-propagating-the-error-to-the-calling-function)?

File: projects/output-and-error-handling/client/main.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+		return retryAfterInteger
+	}
+
+	// third condition a timestamp in string format
+	retryAfterTimeStamp, err := time.Parse(time.RFC1123, retryAfterHeaderString)
+	// if we cannot parse the timestamp then exit
+	if err != nil {
+		fmt.Fprintln(os.Stderr, "Retry-After Header is neither 'a while', an integer or a timestamp")
+		return 0
+	}
```
Comment (berkeli) [inline projects/output-and-error-handling/client/main.go:55]: How can we ensure the caller(caller function, not the caller as a user :) ) knows there was an error? 

How can we ensure the caller and the user know that the error was during `strconv.Atoi` vs `time.Parse`

How can we ensure the returned value `0` is the value we want to return vs `0` (un-initialized int)?

File: projects/output-and-error-handling/client/main.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+	// second condition an integer in string format
+	retryAfterInteger, err := strconv.Atoi(retryAfterHeaderString)
+	// if we successfully parse an integer return it
+	if err == nil {
+		return retryAfterInteger
+	}
+
+	// third condition a timestamp in string format
+	retryAfterTimeStamp, err := time.Parse(time.RFC1123, retryAfterHeaderString)
+	// if we cannot parse the timestamp then exit
```
Comment (berkeli) [inline projects/output-and-error-handling/client/main.go:51]: this comment says it will exit, but doesn't :) (I don't think it should exit here btw)

File: projects/output-and-error-handling/client/main.go [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+	retryAfter := parseRetryAfterHeader(response.Header.Get("Retry-After"))
+
+	switch {
+	case retryAfter > 5:
+		fmt.Fprintln(os.Stderr, "Giving up. We can't get you the Weather.")
+	case retryAfter > 1:
+		fmt.Fprintf(os.Stderr, "Things may be a bit slow because we're doing a retry after %d seconds...\n", retryAfter)
+		time.Sleep(time.Duration(retryAfter))
+		makeGetRequest(url)
+	case retryAfter <= 1 && retryAfter > 0:
```
Comment (berkeli) [inline projects/output-and-error-handling/client/main.go:75]: I think there's only one `int` that can satisfy this condition?

File: projects/output-and-error-handling/client/main.go [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+		fmt.Fprintln(os.Stderr, "Things may be a bit slow because we're doing a retry after 1 second...")
+		time.Sleep(1 * time.Second)
+		makeGetRequest(url)
```
Comment (berkeli) [inline projects/output-and-error-handling/client/main.go:78]: recursions can be a great additions, but they also make the code slightly hard to read/debug/test. Especially if they span across multiple functions. 

Can you think of a way to use error propagation instead?

File: projects/output-and-error-handling/client/main.go [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+}
+
+func main() {
+	// for testing purposes, make an infinite loop
+	for {
+		// make the get request
+		makeGetRequest("http://localhost:8080")
+		// sleep for 1 second
+		time.Sleep(1 * time.Second)
+	}
```
Comment (berkeli) [inline projects/output-and-error-handling/client/main.go:121]: instead of using a for loop here and omitting `os.Exit` which is a requirement in the project description, I'd suggest using a for loop outside of this program. 

hint: you can look into [watch](https://man7.org/linux/man-pages/man1/watch.1.html) command

File: projects/output-and-error-handling/client/main.go [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+// They can all be removed at the end to clean it up
+
+func handleOKResponse(response *http.Response) {
+	// try to read the response body
+	responseBody, err := io.ReadAll(response.Body)
+
+	// use the defer keyword early to close the body regardless
+	defer response.Body.Close()
+
+	if err != nil {
```
Comment (berkeli) [inline projects/output-and-error-handling/client/main.go:24]: In go, best practice is to check for errors right after they are given value. So I'd suggest to move `defer response.Body.Close()` either before you read from body or after you check for errors (preferred)

There's nothing technically wrong with this code, but can you think of why we would need to check for errors before anything else?

**Review by berkeli — COMMENTED** [review]
I've left a few comments, few high level notes:
- You can look into single responsibility principle, e.g. your `handleTooManyRequestsResponse` function handles the response, but does it also do something else? (that wouldn't be expected from reading the function name)
- In go, one of the most common coding patterns is [error propogation](https://github.com/bazmurphy/immersive-go-course/tree/main/projects/output-and-error-handling#1-propagating-the-error-to-the-calling-function). Can you think of reasons why? e.g. why would it be better (or worse) to return an error [here](https://github.com/bazmurphy/immersive-go-course/blob/f68a38183eb5d62caba3d7667aca0fcc49a44d2c/projects/output-and-error-handling/client/main.go#L51) instead of exiting as intended by comments?

**Comment by bazmurphy** [comment]
Thanks @berkeli 

Before I adjust,

Do I understand this correctly, that every function here should do error handling...
In that it should return not only the value, but also an error (if any - otherwise nil)
And if there's an error, return the value as some sort of nil value(?) alongside the error (as the error)

In this way when a function errors, at any part of the call stack, it will send the error back up the chain all the way to the original calling function?
So there is a "trace".. So it is easier to follow it and work out where it originated...

**Comment by berkeli** [comment]
> Thanks @berkeli
> 
> Before I adjust,
> 
> Do I understand this correctly, that every function here should do error handling... In that it should return not only the value, but also an error (if any - otherwise nil) And if there's an error, return the value as some sort of nil value(of its type?) alongside the error (as the error)
> 
> In this way when a function errors, at any part of the call stack, it will send the error back up the chain all the way to the original calling function?
> 
> So that original calling function has a error "trace"(?).. Which makes it is easier to follow and work out where the error originated...

Yes you got that right, this ensures that you don't have some little functions which can cause a panic or exit (imagine a very big project). And also gives the caller an option - do they want to exit? do they want to log a warning but continue with other operations? In your case you will be the caller, so you can log it out and `os.Exit` with the correct code described in project description.

**Comment by bazmurphy** [comment]
Had a go at implementing @berkeli's feedback
Regarding error propagation, and exit codes
It is not ""finished"" but I would rather get more feedback at this stage

Also had a go at adding some tests (not comprehensive)
Since I am going through _**Learn Go with Tests**_ at the moment it makes sense to bring that into this project

File: projects/output-and-error-handling/client/main.go [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+	}
+
+	// use the defer keyword early to close the body regardless
+	defer response.Body.Close()
+
+	return string(responseBody), nil
+}
+
+// returned integer is the retry duration in seconds
+func parseRetryAfterHeader(retryAfterHeaderString string) (int, error) {
```
Comment (berkeli) [inline projects/output-and-error-handling/client/main.go:37]: You make it very clear that this will return `int` duration in seconds and it's perfectly fine, but do you think there's another type that you could return instead of `int`? 

hint: this will save you some conversions

File: projects/output-and-error-handling/client/main.go [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+	//  return 0, nil
+	// }
+
+	// if we successfully parse an integer return it
+	if err == nil {
+		return retryAfterInteger, nil
+	}
+
+	// third condition a timestamp in string format
+	retryAfterTimeStamp, err := time.Parse(time.RFC1123, retryAfterHeaderString)
```
Comment (berkeli) [inline projects/output-and-error-handling/client/main.go:61]: can you look into `time.RFC1123` vs `http.TimeFormat`? are they the same?

File: projects/output-and-error-handling/client/main_test.go [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+
+		result, err := parseResponseBody(response)
+
+		if err != nil {
+			t.Errorf("could not parse body")
+		}
+
+		if result != body {
+			t.Errorf("got %q | want %q", body, result)
+		}
```
Comment (berkeli) [inline projects/output-and-error-handling/client/main_test.go:30]: nit: a more common approach to naming variables intests is `expected, got` or `want, got` as in your error message

File: projects/output-and-error-handling/client/main_test.go [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+
+		want := "failed to read response body : EOF"
+
+		if result != "" {
+			t.Errorf("got %v | want %v", result, "")
+		}
+
+		if err.Error() != want {
+			t.Errorf("got %v | want %v", err, want)
+		}
```
Comment (berkeli) [inline projects/output-and-error-handling/client/main_test.go:47]: nit: I'd structure this slightly differently:
- declare variables in the beginning, i.e. `want` (or `wantErr` so it's very clear)
- for error, do you need to check if it's nil? (or not nil?) what happens if there isn't an error, i.e. it's nil?

File: projects/output-and-error-handling/client/main_test.go [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+		timestamp := futureTime.Format(time.RFC1123)
+
+		retryDuration, err := parseRetryAfterHeader(timestamp)
+
+		// expect back integer 4
+		want := 4
+
+		if retryDuration != want || err != nil {
+			t.Errorf("got %v and %v | want %v and nil", retryDuration, err, want)
+		}
```
Comment (berkeli) [inline projects/output-and-error-handling/client/main_test.go:95]: a better pattern would be to check for these values separately, can you think of why?

File: projects/output-and-error-handling/client/main_test.go [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+		response := &http.Response{Header: http.Header{"Retry-After": []string{"4"}}}
+
+		retryDuration, err := handleTooManyRequestsResponse(response)
+
+		// expect back integer 4
+		want := 4
+
+		if retryDuration != want || err != nil {
+			t.Errorf("got %v and %v | want %v and nil", retryDuration, err, want)
+		}
```
Comment (berkeli) [inline projects/output-and-error-handling/client/main_test.go:111]: same comment as above

File: projects/output-and-error-handling/client/main_test.go [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+
+		// check the retryDuration is 0
+		if retryDuration != 0 {
+			t.Errorf("got retryDuration %v and want %v ", retryDuration, 0)
+		}
+
+		// check the errors match
+		if err.Error() != want.Error() {
+			t.Errorf("got %v | want %v", err, want)
+		}
```
Comment (berkeli) [inline projects/output-and-error-handling/client/main_test.go:130]: nit: this works fine, but you can look into `errors.Is` and `errors.As` comparisons if you've got time. These are a little bit more advanced, so no need to spend too much time on them (there will be plenty of opportunities to learn and experiment)

**Review by berkeli — COMMENTED** [review]
I think there's one major difference in your responseHandler functions, can you think of this difference?

hint: They are not handling the response body the same way

File: projects/output-and-error-handling/client/main.go (hunk 8 above)
Comment (bazmurphy) [reply projects/output-and-error-handling/client/main.go]: I went with time duration as the type because that makes the most sense(?) as in its describing the purpose the value
But i have no idea if you meant that (seems like you were hinting at string?)
As a benefit it also means the retry works in decimal increments

File: projects/output-and-error-handling/client/main.go (hunk 9 above)
Comment (bazmurphy) [reply projects/output-and-error-handling/client/main.go]: I looked into it... `http.TimeFormat` deals with more than just RFC1123 there are apparently multiple RFC specs that HTTP can use in the headers... so using `http.TimeFormat` makes it more robust

File: projects/output-and-error-handling/client/main.go [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+	}
+
+	// if we reach here, we have an retry delay integer
+	switch {
+	case retryDuration > 5:
+		return 0, fmt.Errorf("[2] retry-after is more than 5 seconds")
+	case retryDuration > 1:
+		return retryDuration, nil
+	case retryDuration == 1:
+		return 1, nil
```
[the remark this answers is not in the document]
Comment (bazmurphy) [reply projects/output-and-error-handling/client/main.go]: I went with anything greater than 0, which works well when now using time duration

File: projects/output-and-error-handling/client/main_test.go (hunk 12 above)
Comment (bazmurphy) [reply projects/output-and-error-handling/client/main_test.go]: so that the failure is more granular in that it points to the exact thing that is failing and not leave the developer to guess which of the two it is?

also from this feedback i deliberately put all the err != nil checks first in the test conditions

not sure if that's good or not

**Comment by bazmurphy** [comment]
> I think there's one major difference in your responseHandler functions, can you think of this difference?
> 
> hint: They are not handling the response body the same way

At the moment, I am out of ideas...

If its 200, we want to handle the body etc.
If its 429, there is no body and so no need to even look at it etc.

200 the body is the important part
429 the retry-header is the important part

To me these are two fundamentally different things

Actually the closer the functions get to main the less happy I am with them...

I don't know how to unblock myself... because I need to look at all of this from a totally different viewpoint to find the answer... which usually involves taking a break from it for a bit...

File: projects/output-and-error-handling/client/main.go (hunk 8 above)
Comment (berkeli) [reply projects/output-and-error-handling/client/main.go]: No I meant `time.Duration`

File: projects/output-and-error-handling/client/main_test.go (hunk 11 above)
Comment (berkeli) [reply projects/output-and-error-handling/client/main_test.go]: So in this case if there was an edge case and your result was `""` and your err was `nil` you would have an uncaught panic during test runs (you can't predict how all of the dependecies will behave, so it's always best to avoid relying on "it should or should not do happen")

I'm referring to `err.Error()` which panics if err is nil

File: projects/output-and-error-handling/client/main_test.go (hunk 12 above)
Comment (berkeli) [reply projects/output-and-error-handling/client/main_test.go]: Yeah that would be the right away, it takes mental strain away when you are debugging failing tests.

**Comment by berkeli** [comment]
> @berkeli I have adjusted some of it based on the feedback
> 
> > I think there's one major difference in your responseHandler functions, can you think of this difference?
> > hint: They are not handling the response body the same way
> 
> At this moment, I am out of ideas... which is disappointing because overall this project is ""simple""...
> 
> If its 200, we want to handle the body (the body is the important part) If its 429, there is no body and so no need to even look at it (the retry-header and duration is the important part)
> 
> To me these are two fundamentally different things
> 
> Actually the closer the functions get to main() the less happy I am with them... I feel like it should all be very succint and clean... and I don't feel it is
> 
> I don't know how to unblock myself... because I need to look at all of this from a totally different viewpoint to find the "simple" answer... which usually involves taking a break from it for a bit... but I also want to finish this because it's not _**that**_ far off...
> 
> This feels like one of those "it's this pattern" "yes I didn't think about that but its simple/elegant" times
> 
> And i don't want to look at any model solutions yet because I don't think that is the right way to learn

So the main difference is that you are calling `response.Body.Close()` only in one of the functions, so whenever the request follows a path of `handleTooManyRequestsResponse` you have a resource leak. 

Essentially, when you do an `http.Get()` the http client opens a tcp connection to the destination, when you call `response.Body.Close()` it sends a signal to the client that you are done with that tcp connection and it can be closed. So when you don't call it, it can leave hanging connections and cause a resource leak.

File: projects/output-and-error-handling/client/main_test.go (hunk 11 above)
Comment (bazmurphy) [reply projects/output-and-error-handling/client/main_test.go]: So just the act of trying to "read" `err.Error()` if its `nil` will panic... right?

The thing I am learning here is that Go wants to panic at any and every opportunity, 
so you have to be thorough/clinical in how you handle every single possibility to prevent panics

I think this will come with time as I write more Go, 
I am not comfortable enough yet to read Go and instantly spot what I will term "panic holes" 😆

**Comment by bazmurphy** [comment]
> > @berkeli I have adjusted some of it based on the feedback
> > > I think there's one major difference in your responseHandler functions, can you think of this difference?
> > > hint: They are not handling the response body the same way
> > 
> > 
> > At this moment, I am out of ideas... which is disappointing because overall this project is ""simple""...
> > If its 200, we want to handle the body (the body is the important part) If its 429, there is no body and so no need to even look at it (the retry-header and duration is the important part)
> > To me these are two fundamentally different things
> > Actually the closer the functions get to main() the less happy I am with them... I feel like it should all be very succint and clean... and I don't feel it is
> > I don't know how to unblock myself... because I need to look at all of this from a totally different viewpoint to find the "simple" answer... which usually involves taking a break from it for a bit... but I also want to finish this because it's not _**that**_ far off...
> > This feels like one of those "it's this pattern" "yes I didn't think about that but its simple/elegant" times
> > And i don't want to look at any model solutions yet because I don't think that is the right way to learn
> 
> So the main difference is that you are calling `response.Body.Close()` only in one of the functions, so whenever the request follows a path of `handleTooManyRequestsResponse` you have a resource leak.
> 
> Essentially, when you do an `http.Get()` the http client opens a tcp connection to the destination, when you call `response.Body.Close()` it sends a signal to the client that you are done with that tcp connection and it can be closed. So when you don't call it, it can leave hanging connections and cause a resource leak.
> 
> In your case, I would extract the `defer resource.Body.Close()` to the beginning of `handleStatusCode` so it's called regardless of which path is followed. The body needs to be closed regardless of status code

(immediate reaction without doing any research)
**warning**: _**there are potential misconceptions in what i'm about to write**_
but i will continue anyway to explain my reasoning, and they feel pretty major to me if i haven't understood correctly

I have a lot of questions around this...
I thought an HTTP Response can legitimately be "without a body"
(In that I have written a lot of TS where only the Headers are defined (no body set) eg. in Authorization Request/Responses)

I put the `defer response.Body.Close()` in that function intentionally because 
1. that is the only place we are reading from the body
2. there is no need to close a body on a 429 when never attempt to read the body, we are only reading from its header
3. that is the lowest down the function chain it can go closest to where it's required, so it never even occurred to me that I need to close a body that i am never reading on the 429

So if that is the case then
1. How do we close a body on a response that has no body?
2. Why do we need to close a non-existent body that we never read from?
3. Why is "closing the body" the "trigger" that sends the termination signal... when the response may have no body... that seems strange to me?

I will go try to understand this now

**Comment by bazmurphy** [comment]
_The reason you have to close the response body, even if there is no actual body content, is related to the way Go manages resources and memory. When you make an HTTP request, Go allocates certain resources (e.g., network connections, buffers, etc.) to handle the response. If you don't close the response body, these resources will not be released, leading to potential resource leaks and memory issues, especially in long-running applications or servers that handle a large number of requests._

_Closing the response body is a way to signal to Go that you're done with the response, and it can safely release any associated resources. Even if the response body is empty, there may still be underlying resources that need to be cleaned up.
If you don't close the response body, you may encounter issues such as:_

- _Resource leaks: Unclosed connections, buffers, and other resources will accumulate, eventually leading to resource exhaustion and potential crashes or performance degradation._
- _Memory leaks: Some parts of the response handling code may allocate memory that won't be reclaimed until the response body is closed._
- _Connection issues: Failing to close connections promptly can lead to exhausting the maximum number of open connections allowed by the server or operating system, causing subsequent requests to fail._

_To avoid these issues, it's a best practice to always close the response body after you're done with it, regardless of whether the body contains data or not. Go's `defer` statement is often used to ensure that the response body is closed, even in the presence of errors or early returns._


This gives me my answer... but I still find that strange

**Comment by berkeli** [comment]
> This answers most of my questions above... however i don't think the pattern is that intuitive (if you don't know about it)

It might seem weird, but it does make sense:
- If you make a request and it returns a non-200 status code, it can still have a body (with error details)
- if you want to make a request and don't care about it's body, then you can make a `HEAD` request. This is a method just like `GET`, `PUT`, etc but it will only return the headers and the response doesn't have a body. Not really applicable to this project though

File: projects/output-and-error-handling/client/main.go [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+
+// Comments ----------
+// I write a lot of comments, it has helped me the entire course to learn, understand and remember
+// They can all be removed at the end to clean it up
+
+// Error Messages ----------
+// The Errors are deliberately numbered [x] so I could follow them up the call stack
+// They can be replaced with "user friendly" messages later
+
+func parseResponseBody(response *http.Response) (string, error) {
```
Comment (illicitonion) [inline projects/output-and-error-handling/client/main.go:20]: I'm not sure I'd use "parse" in this function name - this function isn't really _parsing_ the value (taking it in one format an interpreting it in some domain), it's really just reading it (and converting it to a string)?

File: projects/output-and-error-handling/client/main.go [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+	// check if there is a response body
+	if response.Body == nil {
```
Comment (illicitonion) [inline projects/output-and-error-handling/client/main.go:22]: What is this check guarding against? Have you seen a case in real life where this code was needed? (I'm curious why you don't also feel the need to guard against `response` itself being `nil`, but do for `response.Body` which I assume should always be set by an `http.Client`?)

File: projects/output-and-error-handling/client/main.go [hunk 18]
```diff
@@ -0,0 +1,182 @@
...
+
+	return string(responseBody), nil
+}
+
+func parseRetryAfterHeader(retryAfterHeaderString string) (time.Duration, error) {
+	// fmt.Println("--- DEBUG retryAfterHeaderString", retryAfterHeaderString)
+
+	// first condition "a while"
+	if retryAfterHeaderString == "a while" {
+		return 0, fmt.Errorf("[3] retry-after header is 'a while'")
```
Comment (illicitonion) [inline projects/output-and-error-handling/client/main.go:42]: What do you think we should do if the server returns "a while"? It seems like a strange thing to special-case to give this specific error...

File: projects/output-and-error-handling/client/main.go [hunk 19]
```diff
@@ -0,0 +1,182 @@
...
+	if err == nil {
+		// convert the integer into seconds and return it
+		retryDuration := time.Duration(retryAfterInteger) * time.Second
+		return retryDuration, nil
+	}
+
+	// third condition a timestamp in string format
+	// (!) http.ParseTime can parse various formats
+	// specified in the RFCs for HTTP (RFC1123, RFC850, ANSI C etc)
+	// so it is a better choice than time.Parse(RFC1123, _)
```
Comment (illicitonion) [inline projects/output-and-error-handling/client/main.go:58]: This is an interestingly absolute claim :) I don't necessarily disagree with it, but what arguments could you make for choosing `time.Parse(RFC1123, _)` over `http.ParseTime`? What arguments could you make for choosing `http.ParseTime` over `time.Parse(RFC1123, _)`?

File: projects/output-and-error-handling/client/main.go [hunk 20]
```diff
@@ -0,0 +1,182 @@
...
+	retryAfterTimeStamp, err := http.ParseTime(retryAfterHeaderString)
+
+	// if we cannot parse the timestamp then return an error
+	if err != nil {
+		return 0, fmt.Errorf("[3] retry-after header is invalid: %w", err)
```
Comment (illicitonion) [inline projects/output-and-error-handling/client/main.go:63]: It may be useful to include the contents of the header the error message so that someone could debug _why_ it was invalid (and maybe report a bug or fix the server or client)

File: projects/output-and-error-handling/client/main.go [hunk 21]
```diff
@@ -0,0 +1,182 @@
...
+		if err != nil {
+			return "", 0, fmt.Errorf("[1] handleTooManyRequestsResponse failed: %w", err)
+		}
+
+		// return a retry duration
+		return "", retryDuration, nil
+
+	// Other Status Codes
+	default:
+		return "", 0, fmt.Errorf("[1] unhandled response status code: %d", response.StatusCode)
```
Comment (illicitonion) [inline projects/output-and-error-handling/client/main.go:125]: I'd consider including the response body here in case it says something useful, but I think it's also completely fine not to.

File: projects/output-and-error-handling/client/main.go [hunk 22]
```diff
@@ -0,0 +1,182 @@
...
+		return "", 0, fmt.Errorf("[0] request failed: %w", err)
+	}
+
+	// we must Close the body to signal to Go that we are done with the Response
+	// so it can release all the allocated resources
+	// use the defer keyword early to close the body regardless of the outcome
+	defer response.Body.Close()
+
+	// send the response to the status code handler
+	responseBody, retryDuration, err := handleStatusCode(response)
```
Comment (illicitonion) [inline projects/output-and-error-handling/client/main.go:144]: I would consider just `return handleStatusCode(response)` here - I'm not sure how much value all the conditions below are adding - WDYT?

File: projects/output-and-error-handling/client/main.go [hunk 23]
```diff
@@ -0,0 +1,182 @@
...
+		return "", retryDuration, nil
+	}
+
+	// return the parsed response body
+	return responseBody, 0, nil
+}
+
+func main() {
+	// loop until a successful response or an unrecoverable error
+	for {
```
Comment (illicitonion) [inline projects/output-and-error-handling/client/main.go:162]: In the event of retries, this code currently handles each response independently. This is fine, but I'm curious what you think the behaviour should be if the server repeatedly responds saying "sleep for 4 seconds"? What if it does this 100 times in a row?

File: projects/output-and-error-handling/client/main.go [hunk 24]
```diff
@@ -0,0 +1,182 @@
...
+
+		// if we have an unrecoverable error
+		if err != nil {
+			fmt.Fprintln(os.Stderr, err)
+			os.Exit(2)
+		}
+
+		// if we have a retry duration
+		if retryDuration > 0 {
+			fmt.Fprintf(os.Stderr, "Retrying after %v...\n", retryDuration)
```
Comment (illicitonion) [inline projects/output-and-error-handling/client/main.go:173]:
```suggestion
			fmt.Fprintf(os.Stderr, "Will retry after %v...\n", retryDuration)
```

With "Retrying" it's unclear from reading whether the delay has already happened or is about to, whereas "Will retry" makes clear "I'm not attempting the retry yet, I'm sleeping first".

File: projects/output-and-error-handling/client/main_test.go [hunk 25]
```diff
@@ -0,0 +1,177 @@
...
+	"io"
+	"net/http"
+	"testing"
+	"time"
+)
+
+// This is the beginning of the test implementation
+// It is not comprehensive by any means
+
+func TestParseResponseBody(t *testing.T) {
```
Comment (illicitonion) [inline projects/output-and-error-handling/client/main_test.go:16]: These tests are all very reasonable as far as using the standard library goes, but they all feel a little repetitive and verbose in the frequent pattern of:

```go
if want != got {
    t.Errorf("got %q | want %q", got, want")
}
```

I'd encourage you to take a look at [`require`, a library for test assertions](https://pkg.go.dev/github.com/stretchr/testify/require) which I'm a big fan of - it allows you to just write:

```go
require.Equal(t, want, got)
```

which is much less repetitive.

The common functions I tend to use from it are: `require.Equal`, `require.NoError`, `require.ErrorContains`, and `require.Contains`.

File: projects/output-and-error-handling/client/main_test.go (hunk 10 above)
Comment (illicitonion) [reply projects/output-and-error-handling/client/main_test.go]: FWIW the _exact_ naming here tends to be a super language-specific style. I would generally expect to see "expected"/"actual" in Java, and "want"/"got" in Go, for instance :)

File: projects/output-and-error-handling/client/main_test.go (hunk 11 above)
Comment (illicitonion) [reply projects/output-and-error-handling/client/main_test.go]: > i find it harder to read when the want is at the top, but i changed them accordingly

I definitely agree there are trade-offs here. Personally I tend to find super pre-declaring things often a bit harder to follow, but one of the pieces of value it provides is that it shows that the expected value didn't depend on any logic - i.e. there's a constant known value to expect, rather than needing to read the test to understand what the expectation is.

In this particular example, I'd probably inline the string/error into a `require.ErrorContains` or `require.EqualError` or `require.ErrorIs` call rather than having a variable at all, but in other places (e.g. where you're also using `want` as a mock response in a `StringBuffer`), initialising it at the top and re-using it definitely makes sense.

> Go wants to panic at any and every opportunity, so you have to be thorough/clinical in how you handle every single possibility to prevent panics

Yes. This is one of the aspects of Go I don't love compared to some other languages - many things can cause panics _and those are not obvious_.

**Review by illicitonion — COMMENTED** [review]
This is generally looking great, congratulations!

+1 to all of Berkeli's thoughtful comments :)

> In your case, I would extract the `defer resource.Body.Close()` to the beginning of `handleStatusCode` so it's called regardless of which path is followed. The body needs to be closed regardless of status code

They way I'd think about this is that a body is a resource you've been given (even though it may be an empty one). And you need to say when you're done with it. And the most reliable way to make sure you always release a resource is to do make sure you do so close to where you were given it. So the move you did to put the `defer response.Body.Close()` near where you first get the `response` ties those two things together nicely. If you rely on several other functions (which may or may not get called) to do the closing, it's less obviously tied together.

Generally ensuring "this always gets done exactly once" is an interesting problem (and there are many different ways different programming languages support of ensuring this). The "defer close immediately after acquisition" is a fairly reliable way to ensure something happens exactly once. If you had different functions (which you think are mutually exclusive - only one ever gets called) closing the resource, maybe the code gets refactored in the future so that both functions are called and you would try closing the response twice, or a third code-path with an early-return gets added which may lead to you not closing it at all.

---

A couple of questions from the exercise text:

> You should decide whether we should sleep for some amount of time (and if so what) and then retry, or give up. Make sure to write down the reasons for your decision.

Did you write this down somewhere? :)

> All error messages are clear and useful to the user.

How happy are you about this at the moment?

File: projects/output-and-error-handling/client/main.go (hunk 17 above)
Comment (bazmurphy) [reply projects/output-and-error-handling/client/main.go]: i am not familiar enough with how a response is "created" (behind the scenes)

File: projects/output-and-error-handling/client/main.go (hunk 18 above)
Comment (bazmurphy) [reply projects/output-and-error-handling/client/main.go]: i have removed this case, so if it's not a valid integer from string, or timestamp from string it defaults to 3rd case behaviour

File: projects/output-and-error-handling/client/main.go (hunk 19 above)
Comment (bazmurphy) [reply projects/output-and-error-handling/client/main.go]: I don't know I assumed that because we are dealing with an http response that i should use http.ParseTime as it would cover all the possible formats that an http response could have... but maybe that is assumptive

File: projects/output-and-error-handling/client/main.go (hunk 22 above)
Comment (bazmurphy) [reply projects/output-and-error-handling/client/main.go:93]: I didn't even think to do that, i guess looking at so much (basic?) Go code and it all returning the deliberate values as specified by the function definition it never occurred to me to return a function call that fulfills those return types

File: projects/output-and-error-handling/client/main.go (hunk 23 above)
Comment (bazmurphy) [reply projects/output-and-error-handling/client/main.go:114]: I have added some rudimentary retry logic with a max of 3 (which seems to be an agreed standard)

**Comment by bazmurphy** [comment]
I have had a go over on this, trying to address most comments (not all).
The error messages are not necessarily clear/useful/user friendly because at the moment it is more important for me to see the "error propagation in action".
